### PR TITLE
Enable network in container

### PR DESCRIPTION
**Pull request by Sasasu**

### CODE REVIEW

**Comment by Sasasu** [comment]
CI https://extensions.ci.gpdb.pivotal.io/teams/main/pipelines/DEV:SASASU:plcontainer-gpdb-6X-STABLE

File: management/config/plcontainer_configuration.xml [hunk 1]
```diff
@@ -24,6 +24,8 @@
                  When not set, the default CPU share is 1024.
             6.3. "use_container_logging" - set to "yes" or "no" for container logging (not for backend)
                  By default, we set "no".
+            6.4. "enable_network" - set to "yes" or "no", default value is "no".
+                 when "enable_network" is "no", the inside UDF did not have internet access.
```
Comment (beeender) [inline management/config/plcontainer_configuration.xml:28]: `internet access.` not only the internet I think? intranet won't be available as well. So just `network access`?

File: src/plc_configuration.c [hunk 2]
```diff
@@ -253,10 +254,37 @@ static void parse_runtime_configuration(xmlNode *node) {
...
+					value = xmlGetProp(cur_node, (const xmlChar *) "enable_network");
+					if (value != NULL) {
+						validSetting = true;
+						if (strcasecmp((char *) value, "yes") == 0) {
+							conf_entry->enableNetwork = true;
+						} else if (strcasecmp((char *) value, "no") == 0) {
+							conf_entry->enableNetwork = false;
+						} else {
+							plc_elog(ERROR, "SETTING element <enable_network> only accepted \"yes\" or"
+								"\"no\" only, current string is %s", value);
```
Comment (beeender) [inline src/plc_configuration.c:273]: missing space between "or" and "no"
and quotes the `%s`, like `\"yes\"`

File: src/plc_configuration.c [hunk 3]
```diff
@@ -253,10 +254,37 @@ static void parse_runtime_configuration(xmlNode *node) {
...
-					 * this should be set by various backend implementation.
-					 */
+
+					// Enforce to not use network for connection. In the future
+					// this should be set by various backend implementation.
 					conf_entry->useContainerNetwork = false;
+
+					// enable or disable network in container. default is no
+					// will be force yes when use_container_network = yes
+					value = xmlGetProp(cur_node, (const xmlChar *) "enable_network");
```
Comment (beeender) [inline src/plc_configuration.c:264]: This whole function doesn't look neat at all, so many duplicated lines :(
But not the scope for this PR. Just some random complains.

File: src/plc_configuration.c [hunk 4]
```diff
@@ -253,10 +254,37 @@ static void parse_runtime_configuration(xmlNode *node) {
...
+						} else {
+							plc_elog(ERROR, "SETTING element <enable_network> only accepted \"yes\" or"
+								"\"no\" only, current string is %s", value);
+						}
+
+						xmlFree((void *) value);
+						value = NULL;
+
+						// force enable when use_container_network = true
+						if (conf_entry->enableNetwork == false && conf_entry->useContainerNetwork == true) {
```
Comment (beeender) [inline src/plc_configuration.c:280]: is this for the future?? `conf_entry->useContainerNetwork = false;` at line 260

File: src/plc_configuration.c [hunk 5]
```diff
@@ -456,14 +484,16 @@ static void print_runtime_configurations() {
...
 			plc_elog(INFO, "    image = '%s'", conf_entry->image);
+			plc_elog(INFO, "    command = '%s'", conf_entry->command);
 			plc_elog(INFO, "    memory_mb = '%d'", conf_entry->memoryMb);
 			plc_elog(INFO, "    cpu_share = '%d'", conf_entry->cpuShare);
-			plc_elog(INFO, "    use container logging  = '%s'", conf_entry->useContainerLogging ? "yes" : "no");
-			if (conf_entry->useUserControl){
+			// skip conf_entry->useContainerNetwork because it is not user settable.
+			plc_elog(INFO, "    use_container_logging  = '%s'", conf_entry->useContainerLogging ? "yes" : "no");
+			plc_elog(INFO, "    enable_network  = '%s'", conf_entry->enableNetwork ? "yes" : "no");
+			if (conf_entry->useUserControl && conf_entry->roles != NULL){
```
Comment (beeender) [inline src/plc_configuration.c:493]: space between `NULL)` and `{`

**Review by beeender — COMMENTED** [review]
The code looks good in general. We may need to consider the safe and unsafe things which we have discussed. But that could be implemented in another PR.

Also @RMTT needs to review this.

File: management/bin/plcontainer [hunk 6]
```diff
@@ -944,7 +949,7 @@ def main():
 
         if options.subparser_name == "runtime-edit":
             if options.editor is None:
-                options.editor  = "vi"
+                options.editor  = os.getenv("EDITOR") or "vi" # default editer is 'vi'
```
Comment (yihong0618) [inline management/bin/plcontainer:952]: maybe work well too?
```
os.getenv("EDITOR", "vi")
```

**Review by yihong0618 — APPROVED** [review]

File: src/plc_configuration.c (hunk 5 above)
Comment (yihong0618) [reply src/plc_configuration.c:493]: > L

maybe we can bring `diskquota` fmt actions and logic to this repo?

Comment (Sasasu) [reply src/plc_configuration.c:493]: I can't live without clang-format. 😱
I am very careful not to press the 'Format' button.

**Comment by Sasasu** [comment]
for the trusted or untrusted PL problem:

in the very beginning of plcontainer. the network is enabled on purpose.
https://github.com/greenplum-db/plcontainer/blob/22aafb803d62de526ce1aba37b68b60e6f6b87c1/src/plc_docker_api.c#L40-L54

and plcontainer is a trusted PL.

network disabled after https://github.com/greenplum-db/plcontainer/commit/37b2b8875e9cd19f83f73c694d6f9745598af64f

this commit removed the `\"DisableNetwork\": false,\n"`, add `\"NetworkDisabled\": %s,\n"` and setting networkdisabled to true.
the commit wants to support unix domain socket. not related to the network. the disabled whole network looks like a bug.

and some doc copy from pljava

```

In the 1.6.0 default policy, 'trusted' (`java`) code has minimal permissions,
suitable for general computation and interacting with the database, and
'untrusted' (`javaU`) code has only the additional permission to access the
file system. Existing user functions that worked in PL/Java 1.5.x and performed
other actions, such as making network connections, will need the appropriate
permissions (such as `java.net.URLPermission` or `java.net.SocketPermission`)
granted via the policy file.
```

after pljava 1.6 (released on Oct 19, 2020). the trusted also can access network by adding `java.net.SocketPermission`.
so can not access the network is not a feature.

File: src/plc_configuration.c (hunk 3 above)
Comment (Sasasu) [reply src/plc_configuration.c:264]: how about the JSON api like
https://github.com/greenplum-db/plcontainer/blob/2f2dbd572efd0c84aed51da3ee44ad31328459d8/src/plc_docker_api.c#L217-L274

refactor it later!

File: src/plc_configuration.c (hunk 4 above)
Comment (Sasasu) [reply src/plc_configuration.c:280]: yes, for the future.

File: src/plc_configuration.c (hunk 3 above)
Comment (RMTT) [reply src/plc_configuration.c:264]: why not use `json-c``

**Review by RMTT — APPROVED** [review]
LGTM

**Comment by xuebinsu** [comment]
It would be great if we could have some test cases that try to establish network connections.

**Comment by Sasasu** [comment]
> > It would be great if we could have some test cases that try to establish network connections.
> 
> That is a good point.
> Is there any network API we can use in the container python? `httplib`? @Sasasu 

I want just do something like `os.exec("/bin/ping -n 1 127.0.0.1").read()`

i think this feature is not need a rush release, i will merge this pr after adding some basic testcase (and remove the QQQ debug log)

File: src/plc_configuration.c (hunk 3 above)
Comment (Sasasu) [reply src/plc_configuration.c:264]: json-c needs alloc memory and is not simple when serializing object array.
still need something like `__JSON_APPEND_WITH_NAME`.

I think the final json-c code will look similar to this code.

**Pull request merged**